src/ir/auth_logic/ast_construction.cc

```diff
@@ -0,0 +1,153 @@
+//-------------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//-------------------------------------------------------------------------------
+
+#include "src/ir/auth_logic/ast_construction.h"
+
+#include <iostream>
+#include <memory>
+#include <optional>
+#include <stdexcept>
+#include <string>
+#include <utility>
+#include <variant>
+#include <vector>
+
+#include "antlr4-runtime.h"
+#include "src/common/logging/logging.h"
+#include "src/common/utils/map_iter.h"
+#include "src/ir/auth_logic/ast.h"
+#include "src/ir/datalog/program.h"
+#include "src/policy/auth_logic/auth_logic_cc_generator_grammar.inc/AuthLogicLexer.h"
+#include "src/policy/auth_logic/auth_logic_cc_generator_grammar.inc/AuthLogicParser.h"
+#include "src/policy/auth_logic/auth_logic_cc_generator_grammar.inc/AuthLogicVisitor.h"
+
+namespace raksha::ir::auth_logic {
+
+static Principal ConstructPrincipal(
+    AuthLogicParser::PrincipalContext* context) {
+  return Principal(context->ID()->getText());
+}
+
+static datalog::Predicate ConstructPredicate(
+    AuthLogicParser::PredicateContext* context) {
+  std::vector<std::string> args;
+  for (auto& statement : context->pred_arg()) {
+    args.push_back(std::move(statement->ID()->getText()));
+  }
+  datalog::Sign sign = context->NEG() == nullptr ? datalog::Sign::kPositive
+                                                 : datalog::Sign::kNegated;
+  return datalog::Predicate(context->ID()->getText(), args, sign);
+}
+
+template <typename T, typename = std::enable_if_t<std::is_base_of<
+                          AuthLogicParser::VerbphraseContext, T>::value>>
+T* GetVerbPhraseContextAs(AuthLogicParser::VerbphraseContext* context) {
+  return dynamic_cast<T*>(context);
+}
+
+static BaseFact ConstructVerbphrase(
+    Principal left_principal, AuthLogicParser::VerbphraseContext* context) {
+  CHECK(context != nullptr) << "Invalid Verbphrasecontext";
+  if (auto* ctx = GetVerbPhraseContextAs<AuthLogicParser::PredphraseContext>(
+          context)) {  // PredicateContext
+    return BaseFact(
+        Attribute(left_principal, ConstructPredicate(ctx->predicate())));
+  }
+  auto* ctx = GetVerbPhraseContextAs<AuthLogicParser::ActsAsPhraseContext>(
+      context);  // ActsAsPhraseContext
+  return BaseFact(
+      CanActAs(left_principal, ConstructPrincipal(ctx->principal())));
+}
```

Comment:
A few suggestions: 
  - Let us use dynamic_cast directly? 
  -  Also, take a reference in and use `&` to take the address later. That way you can avoid adding a check for context != nullptr`. 
  - Use `CHECK_NOTNULL` at the last use case to make sure we are not dereferencing nullptr.

Other similar functions could use the same pattern.

```suggestion
static BaseFact ConstructVerbphrase(
    Principal left_principal, AuthLogicParser::VerbphraseContext& context) {
  if (auto* ctx = dynamic_cast<AuthLogicParser::PredphraseContext*>(&context)) {
    // PredicateContext
    return BaseFact(
        Attribute(left_principal, ConstructPredicate(ctx->predicate())));
  }
  // ActsAsPhraseContext
  auto* ctx = CHECK_NOTNULL(dynamic_cast<AuthLogicParser::ActsAsPhraseContext*>(&context);
  return BaseFact(
      CanActAs(left_principal, ConstructPrincipal(ctx->principal())));
}
```